### PR TITLE
Microbenchmarks: use existing tests

### LINKED ISSUE #62: Microbenchmarks: Use existing benchmark tests
The existing performance benchmarks are end-to-end. This is really great.

Just like in testing, end-to-end is well supported by testing/benchmarking of smaller units.

Today, we have a few `func BenchFoo(b *testing.B)`  method sprinkled throughout the code base.

It would be awesome if we could pick use these benchmarks and just start running them, collecting numbers and then creating a pretty charts for them.

### PR DESCRIPTION
**Pull request by frouioui**
### What is this?
Hello everyone 👋 This pull request contains two major changes. First, support for microbenchmarks. Second, the initial work on migrating the codebase from Python to Golang.

A new `go` directory is introduced where we can find the micro benchmarking tool, which was initiated in [planetscale/vitess](https://github.com/planetscale/vitess/tree/benchproj) by @systay, this tool is now wrapped in a Cobra CLI.

### How to run?

```shell
# build
go build go/main.go

# get help
./main help

# get help for microbench tool command
./main microbench help

# execute microbench tool
./main microbench run ./go/tools/microbench/integration/... output.txt

# execute microbench tool for another go module
./main microbench run ./vitess ./go/vt/sqlparser/... output_sqlparser_vitess.txt
```

Related to #62.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,11 @@
+module github.com/vitessio/arewefastyet
+
+go 1.16
```
Comment (systay) [inline go.mod:3]: I think this should be safe, but just FYI - Vitess is build on 1.15

Comment (frouioui) [reply go.mod:3]: Should we roll back to 1.15 while https://github.com/vitessio/vitess/issues/7572 is still unresolved?

**Comment by frouioui** [comment]
### Save results to MySQL

Saving micro-benchmarks results to a MySQL database is enabled by giving the correct configuration to the CLI. MySQL configuration should preferably be passed through a file rather than CLI flags. The following elements must be present in the file:

```yaml
db-host: <host>
db-user: <user>
db-password: <password>
db-database: <db>
```

The schema used is the same as the one in `./sql/vitess-benchmark.sql`. Note that if the program runs without that configuration, results will not be saved, they will only be printed to both stdout and file system.

**Review by Akilan1999 — APPROVED** [review]

File: go/mysql/config.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+type ConfigDB struct {
+	Host     string
+	User     string
+	Password string
+	Database string
+}
+
+func (cfg ConfigDB) IsValid() (valid bool) {
+	valid = true
```
Comment (systay) [inline go/mysql/config.go:29]: nit: can be simplified to 
```go
return !(cfg.Database == "" || cfg.Password == "" || cfg.User == "" || cfg.Host == "")
```

File: go/mysql/config_test.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	}{
+		{name: "Valid ConfigDB", config: ConfigDB{Host: "host", User: "user", Password: "password", Database: "database"}, wantValid: true},
+		{name: "Invalid ConfigDB, missing host", config: ConfigDB{User: "user", Password: "password", Database: "database"}, wantValid: false},
+		{name: "Invalid ConfigDB, missing user", config: ConfigDB{Host: "host", Password: "password", Database: "database"}, wantValid: false},
+		{name: "Invalid ConfigDB, missing password", config: ConfigDB{Host: "host", User: "user", Database: "database"}, wantValid: false},
+		{name: "Invalid ConfigDB, missing database", config: ConfigDB{Host: "host", User: "user", Password: "password"}, wantValid: false},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if gotValid := tt.config.IsValid(); gotValid != tt.wantValid {
```
Comment (systay) [inline go/mysql/config_test.go:37]: nit: 
```go
assert.Equal(t, tt.wantValid, gotValid, "valid")
```

File: go/tools/microbench/line_test.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+ * /
+ */
+
+package microbench
+
+import (
+	"testing"
+)
+
+func Test_benchmarkRunLine_applyRegularExpr_checkBenchType(t *testing.T) {
```
Comment (systay) [inline go/tools/microbench/line_test.go:25]: nice test!

File: go/tools/microbench/line_test.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+			gotBenchType := line.benchType
+			if gotBenchType != tt.benchTypeWanted {
+				t.Errorf("line.applyRegularExpr() = %v, want %v", gotBenchType, tt.benchTypeWanted)
+			}
+		})
+	}
+}
+
+func BenchmarkApplyRegularExprRegularBenchmark(b *testing.B) {
```
Comment (systay) [inline go/tools/microbench/line_test.go:87]: ha! benchmark all the things! so meta

File: go/tools/microbench/microbench.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+	command.Dir = rootDir
+	out, err := command.Output()
+
+	if err != nil {
+		return err
+	}
+
+	if b.sql != nil {
+		if err := b.registerToMySQL(b.sql); err != nil {
+			log.Fatal(err)
```
Comment (systay) [inline go/tools/microbench/microbench.go:68]: nit: shouldn't you be returning the error here?

**Review by systay — APPROVED** [review]
nice stuff!
only had minor nitpicks. 🙇

File: go/mysql/config.go (hunk 2 above)
Comment (frouioui) [reply go/mysql/config.go]: Resolved in 04bb6e2

File: go/mysql/config_test.go (hunk 3 above)
Comment (frouioui) [reply go/mysql/config_test.go]: Resolved in 35f162d

File: go/tools/microbench/line_test.go (hunk 5 above)
Comment (frouioui) [reply go/tools/microbench/line_test.go:86]: We want speed! 🐎

File: go/tools/microbench/microbench.go (hunk 6 above)
Comment (frouioui) [reply go/tools/microbench/microbench.go]: I should! Fixed in 710b360

**Comment by frouioui** [comment]
Thank you for the feedback @systay, I corrected the minor nitpicks. Regarding the unit test package, as @fatih recommended us to start doing, I went for [quicktest](https://github.com/frankban/quicktest) instead of testify.

**Pull request merged**